DRACOLoader: Web Worker support

donmccurdy opened the pull request:
Possible solution to https://github.com/mrdoob/three.js/issues/13648 — a drop-in replacement for DRACOLoader, intended for use with GLTFLoader. A glTF file may contain multiple compressed meshes, and this loader allows each mesh to be decompressed by a configurable number of workers, without using the main thread. ~~For simplicity, I omitted some functionality of DRACOLoader that glTF does not use, such as point clouds. The result is 150 lines shorter than DRACOLoader.~~

Usage:

```js 
var loader = new THREE.GLTFLoader();
var decoder = new THREE.DRACOLoader();
decoder.setDecoderPath( '/examples/js/libs/draco' );
loader.setDRACOLoader( decoder );
loadel.load( 'model.glb', (gltf) => ... );
```

~~The alternative would be https://github.com/mrdoob/three.js/issues/13664#issuecomment-422168608, which involves LoaderSupport. That's more code, but also has some additional features. I'm not sure how to weigh those tradeoffs at the moment.~~

donmccurdy committed:
Add GLTFDRACOLoader.

donmccurdy committed:
Clean up.

donmccurdy commented:
/cc @FrankGalligan @kaisalmen

Mugen87 commented:
I personally prefer this approach since it's a very clear and compact implementation.

donmccurdy commented:
@sunag I notice you're using the Draco decoder directly in SEA3DLoader rather than using DRACOLoader — if you'd be interested in using something like this, instead, I'm glad to rename the file as WWDRACOLoader to not be glTF-specific.

donmccurdy commented:
I think there are a couple open things here:

- [ ] Create a new loader vs modifying DRACOLoader?
  - DRACOLoader's `.load()` method wouldn't benefit from this as directly, since each `.drc` file has only a single mesh.
- [ ] Use LoaderSupport, or not?
  - The worker-related complexity seems manageable as-is, but I'm open to discussing this.
- [ ] This implementation currently downloads the draco decoder library to each worker.
  - Presumably the browser's caching prevents that from being a major issue, but it might be better to download once on the main thread anyway.

kaisalmen commented:
`WWDRACOLoader` is a wrapper for existing `DRACOLoader` utilizing `WorkerLoader` (`LoaderSupport` should disappear in the future...) that just can be configured to be used by `GLTFLoader`:
https://rawgit.com/kaisalmen/three.js/WorkerLoader_OBJLoader/examples/webgl_loader_worker_gltf.html

PR for the preview version will come soon. Then, people can compare 😉 
WorkerLoader and its toolbox are overkill for a single loader, but if multiple loaders are able to re-use these functions and help them flourish, then it'll pay off

sunag commented:
> @sunag I notice you're using the Draco decoder directly in SEA3DLoader rather than using DRACOLoader — if you'd be interested in using something like this, instead, I'm glad to rename the file as WWDRACOLoader to not be glTF-specific.

Thank you @donmccurdy. Certainly this could happen in the next updates.

kaisalmen commented:
@donmccurdy nice that the new decoder code has no longer dependencies to three.js! 👍

arpu commented on examples/js/loaders/GLTFDRACOLoader.js:
```diff
@@ -0,0 +1,364 @@
+      var attributeType = self[ attributeTypes[ attributeName ] ];
+      var attributeId = attributeIDs[attributeName];
+      var attribute = decoder.GetAttributeByUniqueId( dracoGeometry, attributeId );
+      geometry.attributes.push( this.decodeAttribute( draco, decoder, dracoGeometry, attributeName, attributeType, attribute ) );
```
this should be 
geometry.attributes.push( decodeAttribute( draco, decoder, dracoGeometry, attributeName, attributeType, attribute ) );

donmccurdy replied:
Fixed, thanks!

arpu commented on examples/js/loaders/GLTFDRACOLoader.js:
```diff
@@ -0,0 +1,364 @@
+
+      try {
+
+        return this.decodeGeometry( draco, decoder, buffer, attributeIDs, attributeTypes );
```
return decodeGeometry( draco, decoder, buffer, attributeIDs, attributeTypes );

donmccurdy replied:
Fixed!

donmccurdy committed:
DRACOLoader2: Initial commit.

donmccurdy committed:
DRACOLoader2: Add credentials header.

donmccurdy committed:
DRACOLoader2: Fix GLTFLoader integration, fix task allocation.

donmccurdy committed:
Remove GLTFDRACOLoader.

donmccurdy commented:
Ok, this PR is in better shape now. I think we can now answer a few of the questions mentioned earlier:

(1) Let's replace DRACOLoader, rather than adding a glTF-specific variant of it.
(2) Let's not use LoaderSupport yet. I'm willing to reconsider this once ES modules are source files perhaps, and there's certainly some potential code overlap between this and BasisTextureLoader.
(3) Issues loading the Draco library in web workers efficiently have been fixed.

There are several breaking changes from the original DRACOLoader:

- Removed TriangleStripDrawMode support.
- Removed quantization support (What was `attribute.isQuantized`? How do I test this?).
- Removed setVerbosity, setDrawMode, and setSkipDequantization methods.
- Changed static initialization methods to instance methods.

None of these affect usage in GLTFLoader, so the new and old versions are both compatible with the existing GLTFLoader. The initialization changes slightly:

```js
// Before
var dracoLoader = new THREE.DRACOLoader();
THREE.DRACOLoader.setDecoderPath( 'js/libs/draco/gltf/' )
loader.setDRACOLoader( dracoLoader );

// After
var dracoLoader = new THREE.DRACOLoader();
dracoLoader.setDecoderPath( 'js/libs/draco/gltf/' )
loader.setDRACOLoader( dracoLoader );
```

donmccurdy commented:
There are two advantages to this approach:

1. Decoding does not block the main thread. This is true regardless of the model, and significant especially for WebXR applications.
2. Decoding is parallelized across workers, and may or may not complete more quickly depending on the file and the device concurrency. For a model with a single mesh, it will be slower, but I consider that an acceptable tradeoff because of (1). For models with many meshes, and especially many _large_ meshes, the load time improvement is significant.

Running tests on two glTF models with multiple meshes (attached below), here are the results:

* **matilda.glb** (103 meshes, 6MB uncompressed geometry)
  - 3% load time decrease, averaged over 5 tests. (527ms -> 512ms)
* **lovecraftian.glb** (18 meshes, 45MB uncompressed geometry)
  - 60% load time decrease, averaged over 5 tests. (3815ms -> 1509ms)

[draco_perf_tests.zip](https://github.com/mrdoob/three.js/files/3308973/draco_perf_tests.zip)

| matilda | lovecraftian |
|---|---|
| ![Screen Shot 2019-06-19 at 11 04 48 PM](https://user-images.githubusercontent.com/1848368/59824677-b0ed8400-92e6-11e9-9d20-43b498cf3abb.png) | ![Screen Shot 2019-06-19 at 11 04 35 PM](https://user-images.githubusercontent.com/1848368/59824678-b2b74780-92e6-11e9-9eb1-ddceb24eff84.png) |

donmccurdy commented:
The file in this PR is named `DRACOLoader2.js`, but I'd like to update DRACOLoader directly if there are no objections to the breaking changes mentioned in https://github.com/mrdoob/three.js/pull/15249#issuecomment-503886776.

donmccurdy committed:
Fix decodeAttribute reference.

donmccurdy committed:
Rename DRACOLoader2 -> DRACOLoader, update example.

donmccurdy committed:
DRACOLoader: Update docs, add missing setPath() method.

donmccurdy committed:
Update webgl_loader_gltf_extensions.html.

donmccurdy committed:
DRACOLoader: Update docs and typings.

donmccurdy committed:
Fix merge error.

donmccurdy commented:
I've added the changes directly to DRACOLoader (not a new loader), and I think this is ready  – here's a demo of my viewer, using the web workers:

https://gltf-viewer-experimental.donmccurdy.com/
